### PR TITLE
:tada: Updates to processing log

### PR DESCRIPTION
**Pull request by Marigold**
Update processing log to handle more cases like renamed variables, etc. It adds a new field `target` that is similar to `variable_name`, but with an additional random suffix that lets us reconstruct the complete computational graph.

I refactored processing log a bit by:
* Creating ProcessingLog and LogEntry objects
* Removed `amend_log` (I didn't find any use for it and thought it'd be nicer if processing log was immutable)
* Moved the logic to `processing_log.add_entry` and `variable.update_log` with sensible defaults

I got it into a state where it's mergeable, though definitely not perfect. I've only polished it on a few examples. There's a lot to do in terms of refactoring and testing, but I think it only makes sense to go all in once we confirm this is useful.

## How to review:

I'd recommend creating a new `etl-plog` directory next to `etl` and experimenting with the processing log there, otherwise you end up with a mix of plog / non-plog datasets which could be confusing (they work together, but it's still cleaner).

1. Check out docs at `docs/architecture/metadata/processing_log.md`

1. Delete all datasets that used old processing log
```
find data/ -name "*.meta.json" -exec grep -l 'processing_log' {} + | xargs -I{} grep -L 'target' {} | xargs -I{} dirname {} | sort -u | xargs -I{} rm -rf {}
```
2. Rerun dataset with processing log enabled
```
PROCESSING_LOG=1 etl uk_egg_statistics --force
```
3. In notebook, visualise the log
```
from owid.catalog import Dataset
from etl.paths import DATA_DIR

ds = Dataset(DATA_DIR / "garden/animal_welfare/2023-08-01/uk_egg_statistics")
tab = ds['uk_egg_statistics']
tab['number_of_eggs_all'].m.processing_log.display(DATA_DIR)
```

## Open questions

To enable plog you have to run it with `PROCESSING_LOG=1 etl ...`, but you cannot simply add it to your `.env` file, because it'd ruin displaying plog from notebooks. It's not as great, but I found no easier way. The nice thing about this is exploring plog for your datasets. You can just run
```
PROCESSING_LOG=1 etl ... --force --only
```
and then check it from the notebook.

## Examples

Here's an artificial example. I've only tried a couple of datasets.

image

### CODE REVIEW

**Comment by Marigold** [comment]
Some random ideas we could talk about today:

1. Processing log wouldn't go further in history than to loading indicator from a dataset (we could reconstruct entire history by prepending processing log of loaded variable)
2. Decorator `@plog("my transformation")` or context managers `with plog("my transformation"): ` for replacing non-intuitive automated logs
3. Code, in general, is a very good high-level abstraction of the process. Processing log feels a bit like going from code to compiled instructions. If I wanted to see what's going on with the data, I'd first try to read comments and then debug it myself line by line in a notebook

**Comment by Marigold** [comment]
Another example:

image

**Comment by Marigold** [comment]
I got it into a state where it's mergeable, though definitely not perfect. I've only polished it on a few examples. There's a lot to do in terms of refactoring and testing, but I think it only makes sense to go all in once we confirm this is useful.

## How to review:

I'd recommend creating a new `etl-plog` directory next to `etl` and experimenting with the processing log there, otherwise you end up with a mix of plog / non-plog datasets which could be confusing (they work together, but it's still cleaner).

1. Delete all datasets that used old processing log
```
find data/ -name "*.meta.json" -exec grep -l 'processing_log' {} + | xargs -I{} grep -L 'target' {} | xargs -I{} dirname {} | sort -u | xargs -I{} rm -rf {}
```
2. Rerun dataset with processing log enabled
```
PROCESSING_LOG=1 etl animal_welfare --force
```
3. In notebook, visualise the log
```
from owid.catalog import Dataset
from etl.paths import DATA_DIR

ds = Dataset(DATA_DIR / "garden/animal_welfare/2023-08-01/uk_egg_statistics")
tab = ds['uk_egg_statistics']
tab['number_of_eggs_all'].m.processing_log.display(output='html')
```

**Comment by pabloarosado** [comment]
Hi @Marigold quick note on:

> Removed amend_log (I didn't find any use for it and thought it'd be nicer if processing log was immutable)

The main use case for `amend_log` was to have an easy way to add a comment to a complicated operation, e.g.

```
tb["something"] = tb.groupby(...).horrible_operation("why_doing_this").amend_log("Explanation of what was done to this indicator here.")
```
But I agree that a decorator on top of the complicated operations would be a better option. It could also avoid adding many entries to the log, to have just one with a reasonable custom message (meanwhile `amend_log` only amends one entry at a time).

**Review by pabloarosado — APPROVED** [review]
Thanks for working on this! I haven't looked deeply into the guts of the code, but I have tried a few things and it makes sense. I think it would be good to merge this as soon as possible to avoid it from stalling (provided that it doesn't affect any other ETL step unless `PROCESSING_LOG=1`). I just have a few comments/questions:
* I see that the `rename` operation is now omitted. I understand why this happens after other operations, but in general, if you explicitly want to rename a variable, that may also be relevant. For example, if you go to the `uk_egg_statistics` garden step and, right before creating a dataset, you do `tb = tb.rename(columns={"number_of_eggs_all": "test"})` (and change the name in the yaml file accordingly), this rename operation doesn't explicitly show up in the diagram. You do see that the `save` entry in the log calls the saved column as `test`. So, I suppose the logic is robust, and maybe we could simply change the mermaid diagram to also show the name of the column stored (`test`).
* I see that `wrap` can be used for complicated processing functions. But is there any way to wrap any line of code, to avoid the processing log from creating many small irrelevant entries? E.g.
```
with squash_log(operation="custom", comment="Blah blah explanation of some complicated processing"):
    tb = tb.something_huge()
    tb = something_else(tb)
```
But if that's not implemented, it doesn't really matter, I was just wondering if this was already a thing.
* Similar to the above: Something we have mentioned in the past is that we could document all relevant operations directly in the log, instead of having comments in the code (of course, by looking at the code, one would also see these explanations). With the previous implementation this was possible by doing `...update_log(comment="...")` or using `amend_log`. But how would that happen now? Ideally, it would be something like:
```
with comment_log("Explanation of operation"):
    tb = tb.something()
```
Is there anything like this? Again, if not, it doesn't matter, it's just nice to have.
* I cloned a new `etl` repos called `etl-plog` as you suggested. But would that also be necessary in the future, every time you want to check the log, or was that just a suggestion for testing this PR?

Feel free to merge if datadiff doesn't show any issue, or the processing doesn't take significantly longer.

File: lib/catalog/tests/test_processing_log.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    t3 = t1 + t2
+    assert t3["a"].metadata.processing_log.as_dict() == [
+        {"variable": "a", "parents": ["a", "a"], "operation": "+", "target": "a#7921731533"}
+    ]
+
+
+@enable_pl
+def test_sum() -> None:
+    random.seed(0)
+    with pl.enable_processing_log():
```
[the remark this answers is not in the document]
Comment (Marigold) [reply lib/catalog/tests/test_processing_log.py:40]: Nope!

File: lib/catalog/tests/test_processing_log.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                "target": "c#1806341205",
+            },
+        ]
+
+
+@enable_pl
+def test_groupby() -> None:
+    t: Table = Table({"a": [1, 2, 3], "b": ["a", "a", "b"]})
+    t.a.title = "A"
+    t.b.title = "B"
```
[the remark this answers is not in the document]
Comment (Marigold) [reply lib/catalog/tests/test_processing_log.py:63]: You're right, removed.

File: etl/steps/data/meadow/irena/2023-06-26/renewable_electricity_capacity.py [hunk 3]
```diff
@@ -51,16 +52,22 @@ def extract_capacity_from_sheet(excel_object: pr.ExcelFile, sheet_name: str) ->
     return tb
 
 
-def extract_capacity_from_all_sheets(snap: Snapshot) -> Table:
-    # Select sheets that contain data (their names are numbers).
-    excel_object = snap.ExcelFile()
-    sheet_names = [sheet for sheet in excel_object.sheet_names if sheet.strip().isdigit()]
-
+@pl.wrap("extract_capacity_from_sheets")
```
[the remark this answers is not in the document]
Comment (Marigold) [reply etl/steps/data/meadow/irena/2023-06-26/renewable_electricity_capacity.py:55]: It is a proof of concept. It only affects plog and not affect dataset result.

File: docs/architecture/metadata/processing_log.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# Processing log
+!!! warning "In progress."
+
+    The processing log is an experimental feature that is not yet fully tested.
+
+
+This log captures every pandas operation in its metadata, allowing users to track the processing history of a dataset. It is particularly useful for visualizing the data pipeline and aiding in debugging processes.
```
[the remark this answers is not in the document]
Comment (Marigold) [reply docs/architecture/metadata/processing_log.md:7]: Sweet, thanks.

**Pull request merged**